Mathlib/Algebra/SkewMonoidAlgebra/Basic.lean

```diff
@@ -0,0 +1,1481 @@
+    (a • b).toFinsupp = a • b.toFinsupp :=
+  rfl
+
+theorem _root_.IsSMulRegular.skewMonoidAlgebra {S : Type*} [Monoid S] [DistribMulAction S k] {a : S}
+    (ha : IsSMulRegular k a) : IsSMulRegular (SkewMonoidAlgebra k G) a
+  | ⟨_x⟩, ⟨_y⟩, h => congr_arg _ <| ha.finsupp (ofFinsupp.inj h)
+
+theorem toFinsupp_injective :
+    Function.Injective (toFinsupp : SkewMonoidAlgebra k G → Finsupp _ _) :=
+  fun ⟨_x⟩ ⟨_y⟩ => congr_arg _
+
+@[simp]
+theorem toFinsupp_inj {a b : SkewMonoidAlgebra k G} : a.toFinsupp = b.toFinsupp ↔ a = b :=
+  toFinsupp_injective.eq_iff
+
+theorem ofFinsupp_injective :
+    Function.Injective (ofFinsupp : Finsupp _ _ → SkewMonoidAlgebra k G) :=
+  fun _x _y => congr_arg toFinsupp
+
+@[simp]
+theorem toFinsupp_eq_zero {a : SkewMonoidAlgebra k G} : a.toFinsupp = 0 ↔ a = 0 := by
+  rw [← toFinsupp_zero, toFinsupp_inj]
+
+/-- A more convenient spelling of `SkewMonoidAlgebra.ofFinsupp.injEq` in terms of `Iff`. -/
+theorem ofFinsupp_inj {a b} : (⟨a⟩ : SkewMonoidAlgebra k G) = ⟨b⟩ ↔ a = b :=
+  iff_of_eq (ofFinsupp.injEq _ _)
+
+@[simp]
+theorem ofFinsupp_eq_zero {a} : (⟨a⟩ : SkewMonoidAlgebra k G) = 0 ↔ a = 0 := by
+  rw [← ofFinsupp_zero, ofFinsupp_inj]
+
+instance inhabited : Inhabited (SkewMonoidAlgebra k G) := ⟨0⟩
+
+instance nontrivial [Nontrivial k] [Nonempty G] :
+    Nontrivial (SkewMonoidAlgebra k G) := Function.Injective.nontrivial ofFinsupp_injective
+
+instance addCommMonoid : AddCommMonoid (SkewMonoidAlgebra k G) :=
+{ Function.Injective.addCommMonoid toFinsupp toFinsupp_injective toFinsupp_zero
+      toFinsupp_add (fun _ _ => toFinsupp_smul _ _) with
+    toAdd  := SkewMonoidAlgebra.add'
+    toZero := SkewMonoidAlgebra.zero
+    nsmul  := (. • .) }
+
+instance isCancelAdd [IsCancelAdd k] : IsCancelAdd (SkewMonoidAlgebra k G) :=
+  Function.Injective.isCancelAdd toFinsupp toFinsupp_injective toFinsupp_add
+
+instance coeFun : CoeFun (SkewMonoidAlgebra k G) fun _ => G → k := ⟨fun s ↦ s.toFinsupp⟩
+
+@[simp, norm_cast]
+theorem coe_smul [SMulZeroClass R k] (b : R) (v : SkewMonoidAlgebra k G) :
+    ⇑(b • v) = b • ⇑v :=
+  rfl
+
+section Support
+
+/-- For `f : SkewMonoidAlgebra k G`, `f.support` is the set of all `a ∈ G` such that
+`f a ≠ 0`. -/
+def support : SkewMonoidAlgebra k G → Finset G
+  | ⟨p⟩ => p.support
+
+@[simp]
+theorem support_ofFinsupp (p) : support (⟨p⟩ : SkewMonoidAlgebra k G) = p.support := by
+  rw [support]
+
+theorem support_toFinsupp (p : SkewMonoidAlgebra k G) : p.toFinsupp.support = p.support := by
+  rw [support]
+
+@[simp]
+theorem support_zero : (0 : SkewMonoidAlgebra k G).support = ∅ := rfl
+
+@[simp]
+theorem support_eq_empty {p} : p.support = ∅ ↔ (p : SkewMonoidAlgebra k G) = 0 := by
+  rcases p with ⟨⟩
+  simp only [support, Finsupp.support_eq_empty, ofFinsupp_eq_zero]
+
+theorem card_support_eq_zero {p} : p.support.card = 0 ↔ (p : SkewMonoidAlgebra k G) = 0 := by
+  simp only [Finset.card_eq_zero, support_eq_empty]
```

Comment:
What's the point of this lemma? It's a straightforward consequence of others

Reply:
This was added to match `Finsupp.card_support_eq_zero` and `Polynomial.card_support_eq_zero`. It is true that the use cases seem to be limited, see [Mathlib/Algebra/Polynomial/EraseLead](https://github.com/leanprover-community/mathlib4/blob/29d0c941b8a42432a63634d6d82ce60faabfe817/Mathlib/Algebra/Polynomial/EraseLead.lean#L90) and [Mathlib/GroupTheory/Perm/Support](https://github.com/leanprover-community/mathlib4/blob/29d0c941b8a42432a63634d6d82ce60faabfe817/Mathlib/GroupTheory/Perm/Support.lean#L571). We can remove it if you deem it superfluous.

Reply:
Yeah, it looks pretty useless

Reply:
I removed it.